### PR TITLE
Using local postgres for pytest

### PR DESCRIPTION
**Pull request by slundqui**
1. Spinning local postgres container for testing (https://github.com/delvtech/elf-simulations/issues/836)
2. Using conftest for test fixtures to avoid importing fixture dependencies in tests
3. Fixing rounding bugs with fixedpoint (https://github.com/delvtech/agent_0/issues/21)

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -33,3 +34,9 @@ def pytest_exception_interact(call):
     @pytest.hookimpl(tryfirst=True)
     def pytest_internalerror(excinfo):
         raise excinfo.value
+
+
+# Importing all fixtures here and defining here
+# This allows for users of fixtures to not have to import all dependency fixtures when running
+# TODO this means pytest can only be ran from this directory
```
Comment (sentilesdal) [inline conftest.py:48]: i hate that this is true :P

File: lib/chainsync/chainsync/test_fixtures/db_session.py [hunk 2]
```diff
@@ -1,20 +1,73 @@
...
+    with DatabaseJanitor(
+        user=postgres_config.POSTGRES_USER,
+        host="localhost",
+        port=postgres_config.POSTGRES_PORT,
+        dbname=postgres_config.POSTGRES_DB,
+        version=12,
+        password=postgres_config.POSTGRES_PASSWORD,
+    ):
+        engine = initialize_engine(postgres_config)
+        yield engine
```
Comment (sentilesdal) [inline lib/chainsync/chainsync/test_fixtures/db_session.py:67]: any cleanup afterwards?

File: lib/chainsync/chainsync/test_fixtures/db_session.py [hunk 3]
```diff
@@ -1,20 +1,73 @@
...
-from chainsync.db.base import Base
-from sqlalchemy import create_engine
+from chainsync import PostgresConfig
+from chainsync.db.base import Base, initialize_engine
+from pytest_postgresql.janitor import DatabaseJanitor
 from sqlalchemy.orm import Session, sessionmaker
 
 
+@pytest.fixture(scope="session")
+def psql_docker() -> Generator[PostgresConfig, Any, Any]:
```
Comment (sentilesdal) [inline lib/chainsync/chainsync/test_fixtures/db_session.py:17]: what are the Any's here?

File: lib/chainsync/chainsync/db/hyperdrive/schema.py [hunk 4]
```diff
@@ -91,7 +91,7 @@ class WalletInfo(Base):
         BigInteger().with_variant(Integer, "sqlite"), primary_key=True, init=False, autoincrement=True
     )
 
-    blockNumber: Mapped[int] = mapped_column(BigInteger, ForeignKey("poolinfo.blockNumber"), index=True)
+    blockNumber: Mapped[int] = mapped_column(BigInteger, index=True)
```
Comment (sentilesdal) [inline lib/chainsync/chainsync/db/hyperdrive/schema.py:90]: 👏

File: tests/system_test.py [hunk 5]
```diff
@@ -135,25 +131,25 @@ def test_bot_to_db(
...
             "positionDuration": 604800,  # 1 week
             "checkpointDuration": 3600,  # 1 hour
             # TODO this is actually inv of solidity time stretch, fix
             "timeStretch": expected_timestretch,
             "governance": deploy_account.address,
             "feeCollector": deploy_account.address,
-            "curveFee": _to_unscaled_decimal(int(0.1e18)),  # 10%
-            "flatFee": _to_unscaled_decimal(int(0.0005e18)),  # 0.05%
-            "governanceFee": _to_unscaled_decimal(int(0.15e18)),  # 15%
+            "curveFee": _to_unscaled_decimal(FixedPoint("0.1").scaled_value),  # 10%
```
Comment (sentilesdal) [inline tests/system_test.py:138]: _to_unscaled_decimal seems silly

Comment (sentilesdal) [reply tests/system_test.py:138]: so you're called _to_unscaled_decimal on a scaled value, why not just covert straight to a string?  isn't there a string method on fp?

Comment (sentilesdal) [reply tests/system_test.py:138]: yeah can't you just do `Decimal(str(FixedPoint("0.1")))`?

File: README.md [hunk 6]
```diff
@@ -67,9 +67,9 @@ Due to a lack of known precision, operations against Python floats are not allow
 However, operations against `int` are allowed.
 In this case, the `int` _argument_ is assumed to be "unscaled", i.e. if you write `int(8) * FixedPoint(8)` we will scale up the first variable return a `FixedPoint` number that represents the float `64.0` in 18-decimal FixedPoint format.
 So in this example the internal representation of that operation would be `64*10**18`.
-If you cast FixedPoint numbers to ints or floats you will get "unscaled" representation, e.g. `float(FixedPoint(8.0)) == 8.0` and `int(FixedPoint(8.528)) == 8`.
+If you cast FixedPoint numbers to ints or floats you will get "unscaled" representation, e.g. `float(FixedPoint("8.0")) == 8.0` and `int(FixedPoint("8.528")) == 8`.
```
Comment (sentilesdal) [inline README.md:70]: perhaps also warn against using  using floats or exponent notation

File: conftest.py [hunk 7]
```diff
@@ -13,6 +13,12 @@
 # instead of allowing pytest to catch the exception and report
 # Based on https://stackoverflow.com/questions/62419998/how-can-i-get-pytest-to-not-catch-exceptions/62563106#62563106
 
+# IMPORTANT NOTE!!!!!
```
Comment (sentilesdal) [inline conftest.py:16]: 😱

**Review by sentilesdal — APPROVED** [review]
only smol things

File: lib/chainsync/chainsync/test_fixtures/db_session.py (hunk 2 above)
Comment (slundqui) [reply lib/chainsync/chainsync/test_fixtures/db_session.py:67]: Cleanup here is done through the exit of the context manager

File: lib/chainsync/chainsync/test_fixtures/db_session.py (hunk 3 above)
Comment (slundqui) [reply lib/chainsync/chainsync/test_fixtures/db_session.py:17]: Great question, I have no idea, this is the type returned from yields

File: tests/system_test.py (hunk 5 above)
Comment (slundqui) [reply tests/system_test.py:138]: I'm using the private function to do the above to keep the code a bit more readable

File: lib/chainsync/chainsync/test_fixtures/db_session.py (hunk 3 above)
Comment (slundqui) [reply lib/chainsync/chainsync/test_fixtures/db_session.py:17]: https://stackoverflow.com/questions/43658999/what-is-the-return-type-hint-of-a-generator-function

Will update to use `Iterator` here instead

File: tests/system_test.py (hunk 5 above)
Comment (slundqui) [reply tests/system_test.py:138]: Ah I see what you mean. Updating _to_unscaled_decimal to take in FixedPoint

File: tests/system_test.py [hunk 8]
```diff
@@ -18,39 +19,20 @@
...
         print(local_hyperdrive_chain)
 
 
-def _to_unscaled_decimal(scaled_value: int) -> Decimal:
-    return Decimal(str(FixedPoint(scaled_value=scaled_value)))
-
-
-def _decimal_almost_equal(a_val: Decimal, b_val: Decimal) -> bool:
-    return abs(a_val - b_val) < 1e-12
+def _to_unscaled_decimal(fp_val: FixedPoint) -> Decimal:
```
Comment (dpaiton) [inline tests/system_test.py:34]: Instead can you add a method to FixedPoint to allow it to be a Decimal?
```python
def __decimal__(self) -> Decimal:
    r"""Cast to Decimal"""
    return Decimal(str(self))
```

Comment (slundqui) [reply tests/system_test.py:34]: Are we still trying to maintain FixedPoint? I thought we were deprecating it in favor of rust

**Review by dpaiton — APPROVED** [review]
Approved, but I would prefer if we added e.g. `Decimal(FixedPoint("5.2"))` support instead of the helper function.

**Pull request merged**